# Introducción

Ya sabemos cómo leer archivos con pandas y cómo realizar una exploración básica del contenido. En esta sesión vamos a aprender algunas técnicas más avanzadas de Exploración de Datos. También vamos a ver los principios de la Limpieza de Datos, que usamos para dejar nuestros conjuntos de datos listos para ser reestructurados, analizados y visualizados.

La Exploración y la Limpieza van totalmente de la mano. No puedes limpiar sin explorar primero, y gran parte de las técnicas de exploración que usamos tienen como objetivo justamente encontrar inconsistencias, errores, redundancias, etc, en nuestro conjunto de datos para poder deshacernos de ellas.

# Objetivos

1. Leer archivos en formato CSV.
1. Utilizar técnicas más avanzadas de exploración de datos.
1. Utilizar la exploración para encontrar inconsistencias, errores y redundancias.
1. Usar algunas técnicas básicas de limpieza de datos.

La sesión pasada aprendimos a leer archivos en formato JSON. El formato JSON, que es muy parecido a los diccionarios de Python, es sólo uno de los tantos formatos con los que nos vamos a topar.

Los CSVs pertenecen a una clase de formatos donde las columnas de nuestra tabla se delimitan usando lo que se llama un separador. CSV significa Comma-Separated Values y como bien imaginarás significa que se una una coma (,) para separar las columnas.

En un archivo de texto donde cada fila de nuestra tabla tiene su propia línea y donde los valores de cada columna se delimitan usando una coma (,). Leer archivos .csv usando pandas es muy fácil. Lo único que tienes que hacer es lo siguiente:

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../Datasets/melbourne_housing-raw.csv',sep=',')
df

Suburb               Address  Rooms Type      Price Method  \
0      Abbotsford         68 Studley St      2    h        NaN     SS   
1      Abbotsford          85 Turner St      2    h  1480000.0      S   
2      Abbotsford       25 Bloomburg St      2    h  1035000.0      S   
3      Abbotsford    18/659 Victoria St      3    u        NaN     VB   
4      Abbotsford          5 Charles St      3    h  1465000.0     SP   
...           ...                   ...    ...  ...        ...    ...   
19735     Windsor       201/152 Peel St      2    u   560000.0     PI   
19736     Wollert       60 Saltlake Bvd      3    h   525300.0      S   
19737  Yarraville           2 Adeney St      2    h   750000.0     SP   
19738  Yarraville       54 Pentland Pde      6    h  2450000.0     VB   
19739  Yarraville  10/127 Somerville Rd      3    t   645000.0     SP   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis   3/09/2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds   4/02/2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
19735  hockingstuart  29/07/2017       4.6    3181.0  ...       1.0  1.0   
19736      Stockdale  29/07/2017      25.5    3750.0  ...       2.0  2.0   
19737  hockingstuart  29/07/2017       6.3    3013.0  ...       2.0  2.0   
19738        Village  29/07/2017       6.3    3013.0  ...       2.0  1.0   
19739            Jas  29/07/2017       6.3    3013.0  ...       1.0  1.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0    1900.00        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0    1900.00        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
19735       NaN         585.0        NaN   Whittlesea -37.67681   145.00323   
19736       NaN         333.0        NaN      Darebin -37.75884   145.00264   
19737    1999.0         199.0     140.00      Darebin -37.75948   144.99615   
19738    2011.0         238.0     118.00         Hume -37.70322   144.88236   
19739    1980.0           0.0      66.32         Hume -37.69815   144.88019   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
19735  Southern Metropolitan        4380.0  
19736  Northern Metropolitan        2940.0  
19737   Western Metropolitan        6543.0  
19738   Western Metropolitan        6543.0  
19739   Western Metropolitan        6543.0  

[19740 rows x 21 columns]

pandas tiene un muy conveniente método llamado read_csv que nos permite leer archivos .csv directamente. Ése mismo método también puede ayudarnos a leer otros formatos con columnas delimitadas por otros separadores. Por ejemplo, podemos leer .tsv ('tab-separated values'), que son archivos donde cada columna está delimitada por un tab (indentación). Sólo basta con llamar el método con el argumento sep=\t

# Análisis Exploratorio de Datos

Muy bien, empecemos nuestra exploración con las técnicas que ya conocemos primero:

In [3]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [4]:
df.head(10)

Suburb             Address  Rooms Type      Price Method  SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS   Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S   Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S   Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB   Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP   Biggin   
5  Abbotsford    40 Federation La      3    h   850000.0     PI   Biggin   
6  Abbotsford         55a Park St      4    h  1600000.0     VB   Nelson   
7  Abbotsford        16 Maugie St      4    h        NaN     SN   Nelson   
8  Abbotsford        53 Turner St      2    h        NaN      S   Biggin   
9  Abbotsford        99 Turner St      2    h        NaN      S  Collins   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
5  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
6  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   
7  6/08/2016       2.5    3067.0  ...       2.0  2.0     400.0         220.0   
8  6/08/2016       2.5    3067.0  ...       1.0  2.0     201.0           NaN   
9  6/08/2016       2.5    3067.0  ...       2.0  1.0     202.0           NaN   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
5        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
6     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   
7     2006.0        Yarra  -37.7965    144.9965  Northern Metropolitan   
8     1900.0        Yarra  -37.7995    144.9974  Northern Metropolitan   
9     1900.0        Yarra  -37.7996    144.9989  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  
5        4019.0  
6        4019.0  
7        4019.0  
8        4019.0  
9        4019.0  

[10 rows x 21 columns]

In [5]:
df.shape

(19740, 21)

In [6]:
df.dtypes

Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

In [7]:
df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19740 entries, 0 to 19739
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         19740 non-null  object 
 1   Address        19740 non-null  object 
 2   Rooms          19740 non-null  int64  
 3   Type           19740 non-null  object 
 4   Price          15396 non-null  float64
 5   Method         19740 non-null  object 
 6   SellerG        19740 non-null  object 
 7   Date           19740 non-null  object 
 8   Distance       19732 non-null  float64
 9   Postcode       19732 non-null  float64
 10  Bedroom2       15327 non-null  float64
 11  Bathroom       15327 non-null  float64
 12  Car            15327 non-null  float64
 13  Landsize       14944 non-null  float64
 14  BuildingArea   8617 non-null   float64
 15  YearBuilt      9351 non-null   float64
 16  CouncilArea    15296 non-null  object 
 17  Lattitude      15448 non-null  float64
 18  Longti

In [9]:
df.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  19740.000000  1.539600e+04  19732.000000  19732.000000  15327.000000   
mean       2.947163  1.054957e+06      9.861509   3106.533904      2.900568   
std        0.981048  6.452557e+05      5.554233     88.429928      1.007491   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.300000e+05      6.100000   3046.000000      2.000000   
50%        3.000000  8.800000e+05      9.200000   3101.000000      3.000000   
75%        4.000000  1.301000e+06     12.600000   3147.000000      3.000000   
max       12.000000  9.000000e+06     47.400000   3978.000000     30.000000   

           Bathroom           Car       Landsize  BuildingArea    YearBuilt  \
count  15327.000000  15327.000000   14944.000000   8617.000000  9351.000000   
mean       1.548509      1.578065     583.917090    196.806973  1874.165655   
std        0.713385      0.972221    3785.423175    561.558007   393.354888   
min        0.000000      0.000000       0.000000      0.000000     1.000000   
25%        1.000000      1.000000     166.000000     94.000000  1930.000000   
50%        1.000000      2.000000     420.000000    132.000000  1965.000000   
75%        2.000000      2.000000     663.000000    199.000000  1997.000000   
max       12.000000     26.000000  433014.000000  40468.000000  2106.000000   

          Lattitude    Longtitude  Propertycount  
count  15448.000000  15448.000000   19732.000000  
mean     -37.811892    144.996632    7509.802706  
std        0.072708      0.096178    4419.678648  
min      -38.182550    144.431620     249.000000  
25%      -37.856860    144.939323    4407.000000  
50%      -37.808965    145.000900    6567.000000  
75%      -37.761897    145.054710   10331.000000  
max      -37.413810    145.526350   21650.000000

Ejemplo de como ponerlo en la base de datos(equipo)

Ok, algunas cosas importantes:

1. El dataset tiene 19740 filas y 21 columnas
1. Podemos observar que la información que tiene el dataset es básicamente cierta descripción de las propiedades, su locación y tipo y fecha de venta.
1. Hay varias columnas float64, rooms es int64, y el resto son object.
1. Tenemos algunos Nans en el dataset.

¿NaNs? ¿Que quiero decir con eso? Bueno, los NaNs son valores Not a Number que básicamente son valores nulos en nuestro dataset. Estos valores pueden causar muchos problemas, ya que son valores nulos en columnas numéricas pero no podemos realizar operaciones matemáticas con ellos (son valores Not a Number, así que las matemáticas están fuera de las posibilidades). En esta sesión vamos a aprender a lidiar con estos valores, pero antes tenemos que aprender algo llamado funciones vectorizadas, que nos ayudará mucho durante esta exploración.

# Aritmética con Series y Funciones vectorizadas

¿Recuerdas nuestras funciones map y filter? A esas funciones les pasábamos nosotros una función y una lista y nos regresaban una lista con los resultados de aplicarle la función a cada uno de los elementos en orden. Las funciones vectorizadas funcionan muy parecido, pero están optimizadas para funcionar con arreglos de pandas y de numpy (si quieres saber más sobre numpy lee esto). Si tomas un arreglo de pandas (es decir, una Serie) y le aplicas una función vectorizada, la función se aplica a todo el arreglo elemento por elemento y te regresa un arreglo del mismo tamaño con el resultado de la aplicación.

In [10]:
serie_1 = pd.Series([1,2,3,4,5])
serie_1

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [11]:
serie_1

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [12]:
serie_1 * 10

0    10
1    20
2    30
3    40
4    50
dtype: int64

In [13]:
serie_1+100/2

0    51.0
1    52.0
2    53.0
3    54.0
4    55.0
dtype: float64

No tenemos que usar map, ni que declarar una función, ni que usar lambda. Basta con una simple operación matemática. pandas además está optimizado para funcionar de esta manera, así que la velocidad de aplicación es mucho mayor que la combinación de map con listas.

Otra manera de aplicar estas transformaciones es usando funciones vectorizadas. Esto lo podemos hacer usando algunas funciones de numpy, por ejemplo. Numpy es otra librería que es muy común entre los científicos de datos. Ofrece muchas herramientas para realizar cálculos numéricos a altas velocidades. No usaremos mucho esta librería en este módulo, pero es importante entender que se pueden utilizar funciones de numpy para aplicar funciones de manera vectorizada a Series de pandas. Para importar numpy hacemos lo siguente:

In [14]:
import numpy as np

In [15]:
np.power(serie_1,3)

0      1
1      8
2     27
3     64
4    125
dtype: int64

In [16]:
np.sqrt(serie_1)

0    1.000000
1    1.414214
2    1.732051
3    2.000000
4    2.236068
dtype: float64

# Agregaciones

Hay una variación de estas funciones vectorizadas llamadas agregaciones (o reducciones) que lo que hacen es tomar un arreglo, atravesarlo "elemento por elemento" y regresar un solo número que es un "resumen" del arreglo. Este "resumen" es justamente la agregación o reducción. Podemos aplicar estas funciones usando numpy o directamente desde una Serie o DataFrame de pandas. Para efectos prácticos, vamos a utilizar nosotros los métodos que vienen integrados directamente en pandas. Por ejemplo, podemos sumar todos los valores de una Serie de esta manera:

In [17]:
serie_1.sum()

15

O podemos contar el número de elementos en una Serie así:

In [18]:
serie_1.count()

5

Podemos obtener el valor más pequeño de la Serie:

In [19]:
serie_1.min()

1

O el valor mas grande

In [20]:
serie_1.max()

5

# Funciones vectorizadas y Agregaciones con DataFrames

Tanto las funciones vectorizadas como las agregaciones pueden ser aplicadas a DataFrames completos.

Vemos primero las agregaciones. al aplicar una agregación a un DataFrame, lo que obtenemos de regreso es el resultado de aplicar la función a cada una de las columnas (que al final de cuentas son Series, ¿lo recuerdas?). Por ejemplo, tenemos este DataFrame:

In [21]:
df2 = pd.DataFrame(
    {
        'precio': [34,54,223,78,56,12,34],
        'cantidad_en_stock':[3,6,10,2,5,45,2],
        'productos_vendidos':[3,45,23,76,24,6,2]
    }, index=['Pokemaster', 'Cegatron', 'Pikame Mucho', 'Lazarillo de Tormes', 'Stevie Wonder', 'Needle', 'El AyMeDuele']
)

In [22]:
df2

precio  cantidad_en_stock  productos_vendidos
Pokemaster               34                  3                   3
Cegatron                 54                  6                  45
Pikame Mucho            223                 10                  23
Lazarillo de Tormes      78                  2                  76
Stevie Wonder            56                  5                  24
Needle                   12                 45                   6
El AyMeDuele             34                  2                   2

In [23]:
df2.shape

(7, 3)

In [24]:
df2.dtypes

precio                int64
cantidad_en_stock     int64
productos_vendidos    int64
dtype: object

In [25]:
df2.describe()

precio  cantidad_en_stock  productos_vendidos
count    7.000000           7.000000            7.000000
mean    70.142857          10.428571           25.571429
std     70.570127          15.501152           27.011461
min     12.000000           2.000000            2.000000
25%     34.000000           2.500000            4.500000
50%     54.000000           5.000000           23.000000
75%     67.000000           8.000000           34.500000
max    223.000000          45.000000           76.000000

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19740 entries, 0 to 19739
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         19740 non-null  object 
 1   Address        19740 non-null  object 
 2   Rooms          19740 non-null  int64  
 3   Type           19740 non-null  object 
 4   Price          15396 non-null  float64
 5   Method         19740 non-null  object 
 6   SellerG        19740 non-null  object 
 7   Date           19740 non-null  object 
 8   Distance       19732 non-null  float64
 9   Postcode       19732 non-null  float64
 10  Bedroom2       15327 non-null  float64
 11  Bathroom       15327 non-null  float64
 12  Car            15327 non-null  float64
 13  Landsize       14944 non-null  float64
 14  BuildingArea   8617 non-null   float64
 15  YearBuilt      9351 non-null   float64
 16  CouncilArea    15296 non-null  object 
 17  Lattitude      15448 non-null  float64
 18  Longti

mira lo que pasa cuando aplicamos la agregacion sum

In [27]:
df2

precio  cantidad_en_stock  productos_vendidos
Pokemaster               34                  3                   3
Cegatron                 54                  6                  45
Pikame Mucho            223                 10                  23
Lazarillo de Tormes      78                  2                  76
Stevie Wonder            56                  5                  24
Needle                   12                 45                   6
El AyMeDuele             34                  2                   2

In [28]:
df2.sum()

precio                491
cantidad_en_stock      73
productos_vendidos    179
dtype: int64

pandas toma cada una de las columnas, suma todos los valores dentro de cada columna y nos regresa el resultado de las sumas en una nueva Serie, donde el índice son los nombres de las columnas en el DataFrame y los valores son las sumas.

In [29]:
df2.min()

precio                12
cantidad_en_stock      2
productos_vendidos     2
dtype: int64

In [30]:
df2.max()

precio                223
cantidad_en_stock      45
productos_vendidos     76
dtype: int64

Ahora veamos cuando aplicmos funciones vecotizadas a nuestro data frame

In [31]:
df2

precio  cantidad_en_stock  productos_vendidos
Pokemaster               34                  3                   3
Cegatron                 54                  6                  45
Pikame Mucho            223                 10                  23
Lazarillo de Tormes      78                  2                  76
Stevie Wonder            56                  5                  24
Needle                   12                 45                   6
El AyMeDuele             34                  2                   2

Multiplicacion

In [32]:
df2 * 100

precio  cantidad_en_stock  productos_vendidos
Pokemaster             3400                300                 300
Cegatron               5400                600                4500
Pikame Mucho          22300               1000                2300
Lazarillo de Tormes    7800                200                7600
Stevie Wonder          5600                500                2400
Needle                 1200               4500                 600
El AyMeDuele           3400                200                 200

Division

In [33]:
df2 / 100

precio  cantidad_en_stock  productos_vendidos
Pokemaster             0.34               0.03                0.03
Cegatron               0.54               0.06                0.45
Pikame Mucho           2.23               0.10                0.23
Lazarillo de Tormes    0.78               0.02                0.76
Stevie Wonder          0.56               0.05                0.24
Needle                 0.12               0.45                0.06
El AyMeDuele           0.34               0.02                0.02

# NaNs
Como ya dijimos, los NaNs (Not a Number) son valores nulos en nuestro conjunto de datos. Son valores que por alguna razón no se encuentran en nuestro dataset. A la hora de coleccionar nuestros datos, al momento de transcribirlos o de almacenarlos, algo pasó que algunos de esos datos faltan en el dataset final. pandas está diseñado para lidiar con estos datos fácilmente.

Veamos un DataFrame con valores NaN (estamos usando el objeto de numpy np.nan para crear valores Nan):

In [34]:
import numpy as np
import pandas as pd

In [35]:
datos= {'precio': [34,54,np.nan,np.nan,56,12,34],
        'cantidad_en_stock':[3,6,np.nan,np.nan,5,45,2],
        'productos_vendidos':[3,45,23,np.nan,24,6,2]
        }
    
indice = [
     'Pokemaster', 'Cegatron', 'Pikame Mucho', 'Lazarillo de Tormes', 'Stevie Wonder', 'Needle', 'El AyMeDuele'
    
]

In [36]:
df_3 = pd.DataFrame(datos, index=indice)

In [37]:
df_3

precio  cantidad_en_stock  productos_vendidos
Pokemaster             34.0                3.0                 3.0
Cegatron               54.0                6.0                45.0
Pikame Mucho            NaN                NaN                23.0
Lazarillo de Tormes     NaN                NaN                 NaN
Stevie Wonder          56.0                5.0                24.0
Needle                 12.0               45.0                 6.0
El AyMeDuele           34.0                2.0                 2.0

Como puedes ver, a este DataFrame le faltan datos. Los valores NaN que contiene son valores nulos, y nos nos dicen nada útil, en realidad. Podríamos asumir cosas acerca de por qué faltan estos datos, pero no serían más que suposiciones.

Ahora, ¿qué pasa con estos datos nulos? Su presencia puede ser problemática para algunos de los análisis que queremos realizar.

En este caso nuestro dataset es muy pequeño y podemos visualizarlo todo de un solo vistazo. Pero vamos a imaginar que nuestro dataset ese mucho más grande y que necesitamos saber si hay NaNs y cuántos. Para lograr esto podemos usar una función vectorizada llamada isna. isna checa cada valor en nuestro DataFrame, lo transforma en True cuando el valor es igual a NaN y a False cuando no lo es: